### PR TITLE
Add support of the command LPOS

### LINKED ISSUE #1528: Add the support of the LPOS command
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.

### Motivation

This issue considers to support the [lpos](https://redis.io/commands/lpos/) command.

### Solution

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by JackTan25**
It closes #1528.

### CODE REVIEW

**Comment by Yangsx-1** [comment]
Could you add some tests?

**Comment by JackTan25** [comment]
> Could you add some tests?

of course

File: src/commands/cmd_list.cc [hunk 1]
```diff
@@ -19,11 +19,11 @@
  */
 
 #include "commander.h"
+#include "commands/command_parser.h"
 #include "error_constants.h"
 #include "event_util.h"
 #include "server/server.h"
 #include "types/redis_list.h"
-
```
Comment (PragmaTwice) [inline src/commands/cmd_list.cc]: Be careful about newlines e.g. please do not delete newlines without reasons.

File: src/commands/cmd_list.cc [hunk 2]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
   bool dst_left_;
 };
 
+class CommandLPOS : public Commander {
```
Comment (PragmaTwice) [inline src/commands/cmd_list.cc]:
```suggestion
class CommandLPos : public Commander {
```

File: src/commands/cmd_list.cc [hunk 3]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
...
+        if (elem == target_) {
+          counter++;
+          if (counter >= rank_) {
+            indexes.push_back(std::to_string(i));
+            if (indexes.size() >= (u_long)count_ && count_ != 0) break;
+          }
+        }
+        if (comparsions >= max_len_ && max_len_ != 0) break;
+      }
+    }
```
Comment (PragmaTwice) [inline src/commands/cmd_list.cc]: Could you move the storage related code to `type/redis_list.h`, like other list commands?

File: src/commands/cmd_list.cc [hunk 4]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
...
+    redis::List list_db(svr->storage, conn->GetNamespace());
+    // get length of the list
+    uint64_t nums = 0;
+    auto s = list_db.Size(args_[1], &nums);
+    if (!s.ok() && !s.IsNotFound()) {
+      return {Status::RedisExecErr, s.ToString()};
+    }
+    // rank_ can be positive ot negative. if it's zero,
+    // it means the user doesn't specify it.
+    if (rank_ == 0) rank_ = 1;
```
Comment (mapleFU) [inline src/commands/cmd_list.cc]: why not make `rank_` default as 1?

File: src/commands/cmd_list.cc (hunk 3 above)
Comment (JackTan25) [reply src/commands/cmd_list.cc]: 👌

File: src/commands/cmd_list.cc [hunk 5]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
...
+    // it means the user doesn't specify it.
+    if (max_len_ == -1) max_len_ = 0;
+
+    // max_len_ == 0 or count_ == 0
+    // means unlimited length.
+
+    // find element
+    std::vector<std::string> indexes;
+    std::string elem = "";
+    auto counter = 0;
```
Comment (mapleFU) [inline src/commands/cmd_list.cc]: Rather than using `auto counter = 0`, would you mind specify the integer types?

File: src/commands/cmd_list.cc [hunk 6]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
...
+    if (!s.ok() && !s.IsNotFound()) {
+      return {Status::RedisExecErr, s.ToString()};
+    }
+    // rank_ can be positive ot negative. if it's zero,
+    // it means the user doesn't specify it.
+    if (rank_ == 0) rank_ = 1;
+    // it means the user doesn't specify it.
+    if (count_ == -1) count_ = 1;
+    // it means the user doesn't specify it.
+    if (max_len_ == -1) max_len_ = 0;
```
Comment (mapleFU) [inline src/commands/cmd_list.cc]: When user doesn't specify it, would you mind handle it in `Parse`, or just set default to these values?

File: src/commands/cmd_list.cc [hunk 7]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
...
+    // find element
+    std::vector<std::string> indexes;
+    std::string elem = "";
+    auto counter = 0;
+    auto comparsions = 0;
+    if (rank_ > 0) {
+      for (uint64_t i = 0; i < nums; i++) {
+        comparsions++;
+        auto s = list_db.Index(args_[1], i, &elem);
+        if (!s.ok() && !s.IsNotFound()) {
```
Comment (mapleFU) [inline src/commands/cmd_list.cc]: Would you mind test the case for `IsNotFound`?

File: src/commands/cmd_list.cc [hunk 8]
```diff
@@ -555,6 +555,101 @@ class CommandLMove : public Commander {
...
+          counter++;
+          if (counter >= rank_) {
+            indexes.push_back(std::to_string(i));
+            if (indexes.size() >= (u_long)count_ && count_ != 0) break;
+          }
+        }
+        if (comparsions >= max_len_ && max_len_ != 0) break;
+      }
+    } else {
+      for (uint64_t i = nums - 1; i >= 0; i--) {
```
Comment (mapleFU) [inline src/commands/cmd_list.cc]: uint64 is not right. it could underflow.

**Comment by Yangsx-1** [comment]
@JackTan25  You can format the code before commit, see details in https://kvrocks.apache.org/community/contributing/

File: src/commands/cmd_list.cc (hunk 6 above)
Comment (JackTan25) [reply src/commands/cmd_list.cc]: I will use default value. Sorry

**Comment by JackTan25** [comment]
> @JackTan25 You can format the code before commit, see details in https://kvrocks.apache.org/community/contributing/

thanks

File: src/types/redis_list.cc [hunk 9]
```diff
@@ -644,4 +644,60 @@ rocksdb::Status List::Trim(const Slice &user_key, int start, int stop) {
   batch->Put(metadata_cf_handle_, ns_key, bytes);
   return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
 }
+
+// like lpos mylist c, the user_key will be mylist.
+rocksdb::Status List::LPos(const Slice &user_key, std::string target_, int rank_, int count_, int max_len_,
```
Comment (mapleFU) [inline src/types/redis_list.cc:649]:
```suggestion
rocksdb::Status List::LPos(const Slice &user_key, std::string target, int rank, int count, int max_len,
```

File: src/types/redis_list.cc [hunk 10]
```diff
@@ -644,4 +644,60 @@ rocksdb::Status List::Trim(const Slice &user_key, int start, int stop) {
...
+                           std::vector<std::string> *indexes) {
+  std::string ns_key;
+  AppendNamespacePrefix(user_key, &ns_key);
+  ListMetadata metadata(false);
+  rocksdb::Status s = GetMetadata(ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
+  indexes->clear();
```
Comment (mapleFU) [inline src/types/redis_list.cc:656]: why clear indexes after return?

File: src/types/redis_list.cc [hunk 11]
```diff
@@ -644,4 +644,60 @@ rocksdb::Status List::Trim(const Slice &user_key, int start, int stop) {
...
 }
+
+// like lpos mylist c, the user_key will be mylist.
+rocksdb::Status List::LPos(const Slice &user_key, std::string target_, int rank_, int count_, int max_len_,
+                           std::vector<std::string> *indexes) {
+  std::string ns_key;
+  AppendNamespacePrefix(user_key, &ns_key);
+  ListMetadata metadata(false);
+  rocksdb::Status s = GetMetadata(ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
```
Comment (mapleFU) [inline src/types/redis_list.cc:655]: Emm, seems that `s.IsNotFound` is handled both internal and outside?

File: src/types/redis_list.cc [hunk 12]
```diff
@@ -644,4 +644,60 @@ rocksdb::Status List::Trim(const Slice &user_key, int start, int stop) {
...
+  indexes->clear();
+  uint64_t nums = 0;
+  s = Size(user_key, &nums);
```
Comment (mapleFU) [inline src/types/redis_list.cc:658]: `size` can be acquired from metadata, seems that it might help

File: src/types/redis_list.cc [hunk 13]
```diff
@@ -644,4 +644,60 @@ rocksdb::Status List::Trim(const Slice &user_key, int start, int stop) {
...
+  // means unlimited length.
+
+  // find element
+  std::string elem = "";
+  int counter = 0;
+  int comparsions = 0;
+  if (rank_ > 0) {
+    for (uint64_t i = 0; i < nums; i++) {
+      comparsions++;
+      auto s = Index(user_key, i, &elem);
```
Comment (mapleFU) [inline src/types/redis_list.cc:672]: Seems that, this function will call `Index` multiple times, and `Index` internal would get metadata, build snapshot and call get. Would you mind implement a `IndexImpl_`, and use it with a consistent snapshot, or just port the logic of `Index` to outside?

**Review by mapleFU — COMMENTED** [review]
Looks better, would mind add some testing for this?

File: src/types/redis_list.cc (hunk 13 above)
Comment (JackTan25) [reply src/types/redis_list.cc:672]: thanks for advice

**Comment by git-hulk** [comment]
Hi @JackTan25, this was done in https://github.com/apache/kvrocks/pull/1681, I will close this PR. Great thanks for your efforts.

**Pull request closed without merging**